usersession/agent/session_agent_test.go

```diff
@@ -103,3 +110,72 @@ func (s *sessionAgentSuite) TestDying(c *C) {
 		c.Error("agent.Dying() channel was not closed when agent stopped")
 	}
 }
+
+func (s *sessionAgentSuite) TestExitOnIdle(c *C) {
+	agent, err := agent.New()
+	c.Assert(err, IsNil)
+	agent.IdleTimeout = 100 * time.Millisecond
+	startTime := time.Now()
+	agent.Start()
+	defer agent.Stop()
+
+	makeRequest := func() {
+		response, err := s.client.Get("http://localhost/v1/session-info")
+		c.Assert(err, IsNil)
+		defer response.Body.Close()
+		c.Check(response.StatusCode, Equals, 200)
+	}
+	makeRequest()
+	time.Sleep(25 * time.Millisecond)
+	makeRequest()
+
+	select {
+	case <-agent.Dying():
+	case <-time.After(2 * time.Second):
+		c.Fatal("agent did not exit after idle timeout expired")
+	}
+	elapsed := time.Now().Sub(startTime)
+	if elapsed < 125*time.Millisecond || elapsed > 175*time.Millisecond {
+		// The idle timeout should have been extended when we
+		// issued a second request after 25ms.
+		c.Errorf("Expected ellaped time close to 125 ms, but got %v", elapsed)
+	}
+}
+
+func (s *sessionAgentSuite) TestPeerCredentialsCheck(c *C) {
+	var logbuf bytes.Buffer
+	log, err := logger.New(&logbuf, logger.DefaultFlags)
+	c.Assert(err, IsNil)
+	logger.SetLogger(log)
+
+	sa, err := agent.New()
+	c.Assert(err, IsNil)
+	sa.Start()
+	defer sa.Stop()
+
+	// Connections from other users are dropped.
+	uid := uint32(syscall.Geteuid())
+	restore := agent.MockUcred(&syscall.Ucred{Uid: uid + 1}, nil)
```

Comment:
you want to call MockUcred once before starting things and change the result of a fixed mock function under some locking. As it is this test is not "go test -race" clean. there might be other issues, the output is fairly noisy atm

Reply:
I've split the test into three, calling MockUcred before the goroutines are spawned.

The other data races are on access to the buffer we are logging to.  This feels a bit like a shortcoming of our `logger.MockLogger` API: while all writes to the logger are protected by `logger.lock`, there is no way for calling code to acquire that mutex when inspecting the contents of the returned log buffer.  If a test involves multiple goroutines creating log messages, there doesn't seem to be a race free way to check it while the goroutines are still running.

Reply:
Would introducing something like this help (haven't tried it)?:

```
diff --git a/logger/logger.go b/logger/logger.go
index 313e5b4dc..2bc631342 100644
--- a/logger/logger.go
+++ b/logger/logger.go
@@ -102,6 +102,15 @@ func MockLogger() (buf *bytes.Buffer, restore func()) {
        }
 }
 
+// WithLoggerLock invokes f with the global logger lock, useful for
+// tests involving goroutines with MockLogger.
+func WithLoggerLock(f func()) {
+       lock.Lock()
+       defer lock.Unlock()
+
+       f()
+}
+
 // SetLogger sets the global logger to the given one
 func SetLogger(l Logger) {
        lock.Lock()
```

Reply:
That seems cleaner than shutting down the session agent before checking the log message.  I've added this helper (along with a short test) to the PR.